examples/isomorphisms.dx

```diff
@@ -0,0 +1,258 @@
+'# Isomorphisms in Dex
+
+'`Iso a b` is the type of an isomorphism between `a` and `b`.
+
+:t MkIso
+> ((a:Type) ?-> (b:Type) ?-> {bwd: b -> a & fwd: a -> b} -> Iso a b)
+
+'This is a normal ADT, and you can construct your own isomorphisms.
+
+cycleThree : Iso (a & b & c) (b & c & a) =
+  MkIso { fwd = \(a, b, c). (b, c, a)
+        , bwd = \(b, c, a). (a, b, c)
+        }
+
+'Isomorphisms can be applied with `appIso`, applied in reverse with `revIso`,
+and flipped with `flipIso`
+
+:p appIso cycleThree (1, 2.0, 3)
+> (2.0, (3, 1))
+
+:p revIso cycleThree (1, 2.0, 3)
+> (3, (1, 2.0))
+
+:p appIso (flipIso cycleThree) (1, 2.0, 3)
+> (3, (1, 2.0))
+
+'They can also be composed with `<>>`:
+
+:p appIso (cycleThree <>> cycleThree) (1, 2.0, 3)
+> (3, (1, 2.0))
+
+:p appIso (cycleThree <>> cycleThree <>> cycleThree) (1, 2.0, 3)
+> (1, (2.0, 3))
+
+'Note that we assume but do not check that the isomorphism is lawful (i.e.
+`appIso iso $ revIso iso x == x` for all `x`, or equivalently
+`iso <>> (flipIso iso) == idIso`).
+
+'In addition, Dex will automatically write some useful isomorphisms for you
+to extract fields from records and variants. There are four syntactic forms
+that produce isos. We will start with the first two:
+- `#x` produces a "lens-like" record accessor `Iso {x:a & ...r} (a & {&...r})`
+- `#?x` produces a "prism-like" variant matcher `Iso {x:a | ...r} (a | {|...r})`
+
+%passes parse
+:t astype (Iso {a:Int & b:Float & c:Unit} _) #b
+> (Iso {a: Int32 & b: Float32 & c: Unit} (Float32 & {a: Int32 & c: Unit}))
+> === parse ===
+> _ans_:() =
+>   astype (Iso {a: Int & b: Float & c: Unit} _)
+>     (MkIso
+>        { bwd = \(x:(), r:()). {b = x, ...r}
+>        , fwd = \{b = x:(), ...r:()}. (,) x r})
+
+%passes parse
+:t astype (Iso {a:Int | b:Float | c:Unit} _) #?b
+> (Iso {a: Int32 | b: Float32 | c: Unit} (Float32 | {a: Int32 | c: Unit}))
+> === parse ===
+> _ans_:() =
+>   astype (Iso {a: Int | b: Float | c: Unit} _)
+>     (MkIso
+>        { bwd = \v:(). case v
+>                    ((Left x:())) -> {| b = x |}
+>                    ((Right r:())) -> {|b| ...r |}
+>                    
+>        , fwd = \v:(). case v
+>                    {| b = x:() |} -> (Left x)
+>                    {|b| ...r:() |} -> (Right r)
+>                    })
+
+'There are also two "zipper" forms, described later on this page:
+- `#&x` produces a "record-zipper" isomorphism
+  ```
+  Iso ({&...l} & {x:a & ...r}) ({x:a & ...l} & {&...r})
+  ```
+- `#|x` produces a "variant-zipper" isomorphism
+  ```
+  Iso ({|...l} | {x:a | ...r}) ({x:a | ...l} | {|...r})
+  ```
+
+'## Record accessors and lens-like helpers
+Record accessor isomorphisms can be passed into the helper function `getAt`:
+
+:t getAt
+> ((a:Type) ?-> (b:Type) ?-> (c:Type) ?-> (Iso a (b & c)) -> a -> b)
+
+:p getAt #foo {foo=1, bar=2.0}
```

Comment:
It would be kind of cool if we could actually have a type class for application, so that we could extend the group of "callable" types. This would let us avoid having helpers like `getAt`, which make accessing fields quite verbose 😕

Reply:
Right, I think `getAt` would be the more usual application. About languages that do it, I think Racket does to some extent? It always desugars the surface syntax into a normalized representation where [each application is annotated with the `#%app` function that you can override](https://docs.racket-lang.org/reference/application.html).

Reply:
So then we would imagine having something like
```
data Callable t = MkCallable (a : Type) (b : Type) (t -> a -> b)

@instance
lensIsoCallable : Callable (Iso a (b & c)) = MkCallable a b getAt

def app : (MkCallable a b doApp : Callable t) ?=> (f : t) (x : a) : b = doApp f x 
```
Which doesn't actually work because you can't unpack class dictionaries inside pi binders. (I feel like I keep running into issues where we want to have types that depend on values or other types in nontrivial ways, for instance getting the `a` and `b` from a `Callable t`. Seems like that could unlock a lot of cool things.)

Reply:
Right, good point about the binder patterns. Unfortunately, making types depend on the result of pattern matching is a little difficult, but I hope we'll be able to get it to work at some point.